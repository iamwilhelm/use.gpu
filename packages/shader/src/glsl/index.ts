import {
  loadModule,
  loadModuleWithCache,
  loadStaticModule,
  bindEntryPoint,
  defineConstants,
  makeModuleCache,

  bundleToAttribute,
  bundleToAttributes,

  glsl, float, int, uint,
} from './shader';

import {
  bindBundle,
  bindModule,
  bindingsToLinks,
  bindingToModule,
  resolveBindings,
} from './bind';

import {
  castTo,
  swizzleTo,
} from './cast';

import {
  chainTo,
} from './chain';

import {
  diffBy,
} from './diff';

import {
  linkBundle,
  linkModule,
  linkCode,
  getPreamble,
  setPreamble,
} from './link';

import {
  makeASTParser,
  compressAST,
  decompressAST,
  rewriteUsingAST,
} from './ast';

import {
  getHash
} from '../util/hash';

import {
  getBundleHash,
  getBundleKey,
} from '../util/bundle';

export {
  loadModule,
  loadModuleWithCache,
  loadStaticModule,
  bindEntryPoint,
  defineConstants,
  makeModuleCache,

  bundleToAttribute,
  bundleToAttributes,

  glsl, float, int, uint,
} from './shader';

export {
  bindBundle,
  bindModule,
  bindingsToLinks,
  bindingToModule,
  resolveBindings,
} from './bind';

export {
  castTo,
  swizzleTo,
} from './cast';

export {
  chainTo,
} from './chain';

export {
  diffBy,
} from './diff';

export {
  linkBundle,
  linkModule,
  linkCode,
  getPreamble,
  setPreamble,
} from './link';

export {
  makeASTParser,
  compressAST,
  decompressAST,
  rewriteUsingAST,
} from './ast';

export {
  getHash
} from '../util/hash';

export {
  getBundleHash,
  getBundleKey,
} from '../util/bundle';

export const GLSLLinker = {
  loadModule,
  loadModuleWithCache,
  loadStaticModule,
  bindEntryPoint,
  bundleToAttribute,
  bundleToAttributes,
  glsl, float, int, uint,

  defineConstants,

  linkBundle,
  linkModule,
  linkCode,
  getPreamble,
  setPreamble,

  bindBundle,
  bindModule,
  bindingsToLinks,
  bindingToModule,
  resolveBindings,

  castTo,
  chainTo,
  diffBy,
  swizzleTo,

  makeASTParser,
  compressAST,
  decompressAST,
  rewriteUsingAST,

  makeModuleCache,

  getHash,
  getBundleHash,
  getBundleKey,
};

export default GLSLLinker;