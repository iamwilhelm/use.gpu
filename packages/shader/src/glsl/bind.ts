import { ShaderModule, DataBinding } from './types';

import { defineConstants } from './shader';
import { makeBindingAccessors, makeUniformBlock } from './gen';
import { makeResolveBindings, namespaceBinding } from '../util/bind';
import { VIRTUAL_BINDGROUP, VOLATILE_BINDGROUP } from './constants';

export { bindBundle, bindModule } from '../util/bind';

const NO_SYMBOLS = [] as any[];

const getVirtualBindGroup = () => VIRTUAL_BINDGROUP;

export const bindingToModule = (
  binding: DataBinding,
): ShaderModule => {
  const {uniform: {name}} = binding;
  const links = makeBindingAccessors([binding], VIRTUAL_BINDGROUP, VOLATILE_BINDGROUP);
  const module = links[name];
  return {...module, entry: name};
}

export const bindingsToLinks = (
  bindings: DataBinding[],
): Record<string, ShaderModule> => {
  return makeBindingAccessors(bindings, VIRTUAL_BINDGROUP, VOLATILE_BINDGROUP);
}

export const resolveBindings = makeResolveBindings(makeUniformBlock, getVirtualBindGroup);
