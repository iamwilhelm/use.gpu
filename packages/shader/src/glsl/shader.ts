import { Tree } from '@lezer/common';
import { ParsedModule, ParsedModuleCache, ShaderDefine } from './types';

import { makeLoadModule, makeLoadModuleWithCache } from '../util/shader';
import { makeBundleToAttribute, makeBundleToAttributes } from '../util/bundle';
import { decompressAST } from '../util/tree';

import { makeASTParser, compressAST } from './ast';
import { toTypeString, toTypeArgs } from './type';
import { parser } from './grammar/glsl';
import LRU from 'lru-cache';

export { loadStaticModule, loadVirtualModule } from '../util/shader';

// LRU cache for parsed shader code
export const makeModuleCache = (options: Record<string, any> = {}) => new LRU<string, ParsedModule>({
  max: 100,
  ...options,
});

export const DEFAULT_CACHE = makeModuleCache();

// Parse GLSL code into lezer tree
export const parseShader = (code: string): Tree => parser.parse(code);

// Parse a code module into its in-memory representation
// (AST + symbol table)
export const loadModule = makeLoadModule(parseShader, makeASTParser, compressAST);

// Use cache to load modules
export const loadModuleWithCache = makeLoadModuleWithCache(loadModule, DEFAULT_CACHE);

// Make GLSL definitions
export const defineConstants = (defs: Record<string, ShaderDefine>): string => {
  const out = [];
  for (let k in defs) if (defs[k] !== false && defs[k] !== null) out.push(`#define ${k} ${defs[k]}`);
  return out.join("\n");
}

export const bundleToAttribute = makeBundleToAttribute(toTypeString, toTypeArgs);
export const bundleToAttributes = makeBundleToAttributes(toTypeString, toTypeArgs);
