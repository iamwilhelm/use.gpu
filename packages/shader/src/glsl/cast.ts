import { UniformAttribute, ShaderModule, ParsedBundle } from './types';
import { loadVirtualModule } from './shader';
import { getProgramHash } from '../util/hash';
import { toBundle } from '../util/bundle';
import { PREFIX_CAST } from '../constants';

const NO_SYMBOLS = [] as string[];

const arg = (x: number) => String.fromCharCode(97 + x);

export const castTo = (
  bundle: ShaderModule,
  type: string,
  swizzle: string,
): ParsedBundle => {
  const {module, virtual} = toBundle(bundle);
  const {name, format, args} = bundleToAttribute(bundle);

  const {table: {hash}} = module;
  const rehash = getProgramHash(`#cast [${name} ${format}] ${hash}`);
  const symbols = ['cast', 'getValue'];
  const namespace = `${PREFIX_CAST}${rehash.slice(0, 6)}`;

  const imported = {
    at: 0,
    symbols: NO_SYMBOLS,
    name: namespace,
    imports: [{name: 'getValue', imported: name}],
  };
  const modules = [imported];

  // Code generator
  const render = (namespace: string, rename: Map<string, string>) => {
    const name = rename.get('cast') ?? 'cast';
    const accessor = rename.get('getValue') ?? 'getValue';
    return makeCastTo(name, accessor, args ?? [], format, type, swizzle);
  }

  const cast = loadVirtualModule(
    { render },
    { symbols, modules },
    'cast',
    rehash,
  );

  const revirtual = module.virtual
    ? (virtual ? [...virtual, module] : [module])
    : virtual;

  return {module: cast, libs: {[namespace]: bundle}, virtual: revirtual};
}

export const makeCastTo = (
  name: string,
  accessor: string,
  args: string[],
  from: string,
  to: string,
  swizzle: string,
) => {
  const symbols = args.map((t, i) => `${arg(i)}`);
  return `${to} ${name}(${symbols.map((s, i) => `${args[i]} ${s}`).join(', ')}) {
  return ${accessor}(${symbols.join(', ')}).${swizzle};
}
`;
}

export const bundleToAttribute = (
  bundle: ShaderModule,
  name?: string,
): UniformAttribute => {
  // @ts-ignore
  const module = bundle.module ?? bundle;
  const {table: {functions}} = module;

  const entry = bundle.entry ?? module.entry ?? name;

  for (const fn of functions) {
    const {prototype} = fn;
    if (prototype) {
      const {type, name, parameters} = prototype;
      if (name === entry) {
        return {name, format: type.name, args: parameters};
      }
    }
  }

  return {name: name ?? 'main', format: 'void', args: []};
}
