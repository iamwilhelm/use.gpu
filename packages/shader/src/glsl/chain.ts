import { makeChainTo } from '../util/chain';
import { bundleToAttribute } from './shader';

const arg = (x: number) => String.fromCharCode(97 + x);

export const makeChainAccessor = (
  type: TypeLike,
  name: string,
  args: string[],
  from: string,
  to: string,
) => {
  const symbols = args.map((t, i) => `${arg(i)}`);

  return `${type} ${name}(${symbols.map((s, i) => `${args[i]} ${s}`).join(', ')}) {
  return ${to}(${from}(${symbols.join(', ')}));
}
`;   
}

export const chainTo = makeChainTo(makeChainAccessor, bundleToAttribute);
