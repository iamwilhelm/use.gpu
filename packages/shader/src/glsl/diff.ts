import { makeDiffBy } from '../util/diff';
import { bundleToAttribute } from './shader';

const arg = (x: number) => String.fromCharCode(97 + x);

const literal = (v: string | number, isFloat: boolean) => {
  if (typeof v === 'string') return v;
  if (isFloat) {
    let s = v.toString();
    if (!s.match(/[.eE]/)) s = s + '.0';
    return s;
  }
  return Math.round(v).toString();
};

export const makeDiffAccessor = (
  name: string,
  accessor: string,
  args: string[],
  type: string,
  offsets: (number | string)[],
) => {
  const isFloat = (type: string) => !!type.match(/^(float|vec2|vec3|vec4|double|dvec2|dvec3|dvec4)$/);

  const symbols = args.map((t, i) => `${arg(i)}`);

  return `${type} ${name}(${symbols.map((s, i) => `${s}: ${args[i]}`).join(', ')}) {
  ${symbols.map((s, i) => `let d${s} = ${s} + ${literal(offsets[i] ?? 0, isFloat(args[i]))};`)}
  let v1 = ${accessor}(${symbols.map(s => s).join(', ')});
  let v2 = ${accessor}(${symbols.map(s => 'd' + s).join(', ')});
  return v2 - v1;
}
`;
}

export const diffBy = makeDiffBy(makeDiffAccessor, bundleToAttribute);
