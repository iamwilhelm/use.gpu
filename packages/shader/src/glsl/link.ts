import { Tree } from '@lezer/common';
import { ShaderDefine, SymbolTable, ParsedModule, ParsedModuleCache, ParsedBundle, RefFlags as RF, DataBinding } from '../types';
import { parseShader, defineConstants, loadModule, loadModuleWithCache } from './shader';
import { rewriteUsingAST, resolveShakeOps } from './ast';
import { parseBundle, parseLinkAliases } from '../util/bundle';
import { getGraphOrder } from '../util/tree';
import { GLSL_VERSION } from '../constants';
import * as T from '../grammar/glsl.terms';
import mapValues from 'lodash/mapValues';
import { DEFAULT_CACHE } from './shader';

const TIMED = false;

const timed = (name: string, f: any) => {
  if (!TIMED) return f;
  return (...args: any[]) => {
    const t = +new Date();
    const v = f(...args);
    console.log(name, (+new Date() - t), 'ms');
    return v;
  }
}

// Override GLSL version/prefix
let PREAMBLE = `#version ${GLSL_VERSION}`;
export const setPreamble = (s: string): string => PREAMBLE = s;
export const getPreamble = (): string => PREAMBLE;

// Link a source module with static modules and dynamic links.
export const linkCode = timed('linkCode', (
  code: string,
  libraries: Record<string, string> = {},
  links: Record<string, string> = {},
  defines: Record<string, ShaderDefine> = {},
  cache: ParsedModuleCache | null = DEFAULT_CACHE,
) => {
  const main = loadModuleWithCache(code, 'main', undefined, cache);

  const parsedLibraries = mapValues(libraries, (code: string, name: string) => loadModuleWithCache(code, name, undefined, cache));
  const parsedLinkDefs = mapValues(links, (code: string, name: string) => loadModuleWithCache(code, name.split(':')[0], undefined, cache));

  return linkModule(main, parsedLibraries, parsedLinkDefs, defines);
});

// Link a bundle of parsed module + libs and dynamic links.
export const linkBundle = timed('linkBundle', (
  bundle: ParsedBundle,
  links: Record<string, ParsedBundle | ParsedModule> = {},
  defines: Record<string, ShaderDefine> = {},
) => {
  const {code} = linkBundleVirtual(bundle, links, defines);
  return code;
});

// Link a parsed module with static modules and dynamic links.
export const linkModule = timed('linkModule', (
  main: ParsedModule,
  libraries: Record<string, ParsedModule> = {},
  linkDefs: Record<string, ParsedModule> = {},
  defines: Record<string, ShaderDefine> = {},
) => {
  const {code} = linkModuleVirtual(main, libraries, linkDefs, defines);
  return code;
});

// Link a bundle of parsed module + libs, dynamic links and virtual bindings.
export const linkBundleVirtual = timed('linkBundleVirtual', (
  bundle: ParsedBundle,
  links: Record<string, ParsedBundle | ParsedModule> = {},
  defines: Record<string, ShaderDefine> = {},
) => {
  let [main, libs] = parseBundle(bundle);

  const linkDefs = mapValues(links, (bundle: ParsedBundle | ParsedModule, name: string) => {
    if ('table' in bundle) return bundle as any as ParsedModule;

    const [link, linkLibs] = parseBundle(bundle as any as ParsedBundle);
    libs = {...libs, ...linkLibs};
    return link;
  });

  return linkModuleVirtual(main, libs, linkDefs, defines);
});

// Link a parsed module with static modules, dynamic links and virtual bindings.
export const linkModuleVirtual = timed('linkModuleVirtual', (
  main: ParsedModule,
  libraries: Record<string, ParsedModule> = {},
  linkDefs: Record<string, ParsedModule> = {},
  defines: Record<string, ShaderDefine> = {},
) => {
  if (typeof main === 'string') throw new Error("Module is a string instead of an object");

  const [links, aliases] = parseLinkAliases(linkDefs);
  const {modules, exported} = loadModulesInOrder(main, libraries, links, aliases);

  const program = [getPreamble(), defineConstants(defines)] as string[];

  const hashes = new Map<string, string>();
  const namespaces = new Map<string, string>();
  const used = new Set<string>();
  const exists = new Set<string>();
  const visible = new Set<string>();
  const fixed = new Map<string, string>();

  let bindingIndex = 0;
  const nextBinding = () => bindingIndex++;
  const bindings = [] as DataBinding[][];

  for (const module of modules) {
    const {name, code, tree, table, shake, virtual} = module;
    const {hash, globals, symbols, visibles, externals, modules} = table;

    // Multiple links into same module with different name
    if (hashes.has(hash)) {
      namespaces.set(name, hashes.get(hash)!);
      continue;
    }

    // Namespace all non-global symbols outside main module
    const rename = new Map<string, string>();
    if (module !== main) {
      const namespace = reserveNamespace(module, namespaces, used);
      hashes.set(hash, namespace);

      for (const name of symbols) rename.set(name, namespace + name);
      for (const name of globals) rename.set(name, name);

      for (const name of visibles) visible.add(rename.get(name)!);
      for (const name of rename.values()) exists.add(name);

      for (const name of globals) fixed.set(namespace + name, name);
    }

    // Replace imported symbol names with target
    for (const {name: module, imports} of modules) {
      const namespace = namespaces.get(module);
      for (const {name, imported} of imports) {
        let imp = namespace + imported;
        if (fixed.has(imp)) imp = fixed.get(imp)!;
        if (!exists.has(imp)) console.warn(`Import ${name} from '${module}' does not exist`);
        else if (!visible.has(imp)) console.warn(`Import ${name} from '${module}' is private`);
        rename.set(name, imp);
      }
    }

    // Replace imported function prototype names with target
    for (const {flags, prototype} of externals) if (prototype) {
      const {name} = prototype;
      const resolved = aliases?.get(name) ?? name;
      const namespace = namespaces.get(name);

      if ((namespace === undefined) && (flags & RF.Optional)) continue;

      let imp = namespace + resolved;
      if (fixed.has(imp)) imp = fixed.get(imp)!;
      if (!exists.has(imp)) console.warn(`Link ${name}:${resolved} does not exist`);
      else if (!visible.has(imp)) console.warn(`Link ${name}:${resolved} is private`);
      rename.set(name, imp);
    }

    if (virtual) {
      // Emit virtual module in target namespace,
      // with dynamically assigned binding slots.
      const namespace = hashes.get(hash)!;
      const {code, virtuals} = virtual.render(namespace, nextBinding);

      program.push(code);
      bindings.push(...virtuals);
    }
    else {
      // Shake tree ops based on which symbols were exported
      const keep = exported.get(hash);
      const ops = shake && keep ? resolveShakeOps(shake, keep) : null;

      // Rename symbols using AST while tree shaking
      const recode = rewriteUsingAST(code, tree, rename, ops);
      program.push(recode);
    }
  }

  const code = program.join("\n");
  return {code, bindings};
});

// Load all modules references from a piece of code,
// while gathering info about what's exported (for tree shaking).
export const loadModulesInOrder = timed('loadModules', (
  main: ParsedModule,
  libraries: Record<string, ParsedModule>,
  links: Record<string, ParsedModule>,
  aliases: Map<string, string> | null,
) => {
  const graph = new Map<string, string[]>();
  const seen = new Set<string>();
  const hoist = new Set<string>();
  const exported = new Map<string, Set<string>>();

  const out = [] as ParsedModule[];

  // Traverse graph starting from main
  const queue = [{name: main.name, module: main}];
  seen.add(main.name);

  while (queue.length) {
    const next = queue.shift()!;
    const {name, module} = next;
    if (module == null) throw new Error(`Shader module ${name} undefined`);

    out.push(module);

    const {table: {modules, externals}} = module;
    const deps = [] as string[];

    // Recurse into imports
    for (const {at, name, imports} of modules) {
      const isLink = name.match(/^#/);
      const module = isLink ? links[name] : libraries[name];
      if (!module) throw new Error(`Unknown module '${name}'`);

      if (!seen.has(name)) queue.push({name, module: {...module, name}});
      seen.add(name);
      deps.push(name);

      if (at < 0) hoist.add(name);

      const {table: {hash}} = module;
      let list = exported.get(hash);
      if (!list) exported.set(hash, list = new Set());
      imports.forEach(i => list!.add(module.entry ?? i.name));
    }

    // Recurse into links
    for (const {prototype, flags} of externals) if (prototype) {
      const {name} = prototype;
      const module = links[name];
      if (!module) {
        if (flags & RF.Optional) continue;
        throw new Error(`Unlinked function '${name}'`);
      }

      if (!seen.has(name)) queue.push({name, module: {...module, name}});
      seen.add(name);
      deps.push(name);

      const {table: {hash}} = module;
      let list = exported.get(hash);
      if (!list) exported.set(hash, list = new Set());
      list.add(module.entry ?? aliases?.get(name) ?? name);
    }

    // Build module-to-module dependency graph
    graph.set(name, deps);
  }

  // Sort by graph depth
  const order = getGraphOrder(graph, main.name);
  for (let [k, v] of order.entries()) if (hoist.has(k)) order.set(k, v + 1e5);
  out.sort((a, b) => order.get(b.name)! - order.get(a.name)! || a.name.localeCompare(b.name));

  return {
    modules: out,
    exported,
  };
});

// Reserve a unique namespace based on truncated shader hash
export const reserveNamespace = (
  module: ParsedModule,
  namespaces: Map<string, string>,
  used: Set<string>,
) => {
  const {name, table: {hash}} = module;
  let namespace;
  let n = 2;

  do {
    namespace = '_' + hash.slice(0, n++) + '_';
  } while (used.has(namespace));

  namespaces.set(name, namespace);
  used.add(namespace);

  return namespace;
}
