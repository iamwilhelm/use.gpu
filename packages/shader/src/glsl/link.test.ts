import { GLSLModules } from '@use-gpu/glsl';
import { linkCode, linkModule } from './link';
import { loadModule } from './shader';
import { addASTSerializer } from '../test/snapshot';

addASTSerializer(expect);

describe("link", () => {
  
  it("links an external", () => {
    
    const code = `
    vec4 getColor();
    void main() {
      vec4 v;
      v.xyz = vec3(1.0, 0.0, 0.0);
      gl_FragColor = getColor();
    }
    `
    
    const getColor = `
    #pragma export
    vec4 getColor() { return vec4(1.0, 0.0, 1.0, 1.0); }
    `
    
    const linked = linkCode(code, {}, {getColor});
    expect(linked).toMatchSnapshot();

  });

  it("links quad vertex", () => {

    const getPosition = `
    #pragma export
    vec4 getPosition(int index) { return vec4(1.0, 0.0, 1.0, 1.0); }
    `

    const getColor = `
    #pragma export
    vec4 getColor(int index) { return vec4(1.0, 0.0, 1.0, 1.0); }
    `

    const getSize = `
    #pragma export
    float getSize(int index) { return 1.0; }
    `

    const code = GLSLModules['instance/vertex/quad'];
    const modules = GLSLModules;
    const linked = linkCode(code, modules, {getPosition, getColor, getSize});
    expect(linked).toMatchSnapshot();

  });
  
  it("lifts recursive dependency", () => {
    
    const code = `
    #pragma import {getLifted} from 'getLifted'
    #pragma import {getColor1} from 'getColor1'
    void main() {
      gl_FragColor = getColor1();
    }
    `

    const getLifted = `
    // Lifted Code
    #pragma export
    void getLifted() {}
    `

    const getColor1 = `
    #pragma import {getColor2} from 'getColor2'
    #pragma export
    vec4 getColor1() { return getColor2(); }
    `
    
    const getColor2 = `
    #pragma import {getLifted} from 'getLifted'
    #pragma export
    vec4 getColor2() { return vec4(1.0, 0.0, 1.0, 1.0); }
    `
    
    const linked = linkCode(code, {getColor1, getColor2, getLifted});
    expect(linked.indexOf('// Lifted Code')).toBeLessThan(linked.indexOf('getColor2'));
    expect(linked).toMatchSnapshot();

  });
  
  it("links same module twice with different entry point", () => {

    const sub = `
    #pragma export
    vec4 getPosition(int index) { return vec4(1.0, 0.0, 1.0, 1.0); }

    #pragma export
    vec4 getColor(int index) { return vec4(1.0, 0.0, 1.0, 1.0); }
    `

    const main = `
    vec4 getPosition(int index);
    vec4 getColor(int index);
    void main() {
      vec4 a = getPosition(0);
      vec4 b = getColor(0);
    }
    `

    const modMain = loadModule(main, 'main');
    const modSub = loadModule(sub, 'sub');

    const getPosition = {...modSub, entry: 'getPosition'};
    const getColor = {...modSub, entry: 'getColor'};

    const linked = linkModule(modMain, {}, {getPosition, getColor});
    expect(linked).toMatchSnapshot();

  });
});