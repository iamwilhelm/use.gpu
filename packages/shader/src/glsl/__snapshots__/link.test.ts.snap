// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`link lifts recursive dependency 1`] = `
"#version 450
//// @link getLifted _01_


    // Lifted Code
    void _01_getLifted() {}

//// @link getColor2 _02_


    vec4 _02_getColor2() { return vec4(1.0, 0.0, 1.0, 1.0); }

//// @link getColor1 _03_


    vec4 _03_getColor1() { return _02_getColor2(); }

//// @link main


    void main() {
      gl_FragColor = _03_getColor1();
    }
    "
`;

exports[`link links a global across a module 1`] = `
"#version 450
//// @link sub1 _01_


    vec4 getPosition(int index) { return vec4(1.0, 0.0, 1.0, 1.0); }

//// @link sub2 _02_


    void _02_getPosition() {};

    vec4 _02_getColor(int index) {
      _02_getPosition();
      return vec4(1.0, 0.0, 1.0, 1.0);
    }

//// @link main


    vec4 getPosition(int index);
    void main() {
      vec4 a = getPosition(0);
    }
    "
`;

exports[`link links an external 1`] = `
"#version 450
//// @link getColor _01_


    vec4 _01_getColor() { return vec4(1.0, 0.0, 1.0, 1.0); }

//// @link main


    vec4 _01_getColor();
    void main() {
      vec4 v;
      v.xyz = vec3(1.0, 0.0, 0.0);
      gl_FragColor = _01_getColor();
    }
    "
`;

exports[`link links quad vertex 1`] = `
"#version 450
//// @link main _01_


    vec4 _01_getPosition(int index) { return vec4(1.0, 0.0, 1.0, 1.0); }

//// @link main _02_


    vec4 _02_getColor(int index) { return vec4(1.0, 0.0, 1.0, 1.0); }

//// @link main _03_


    float _03_getSize(int index) { return 1.0; }

//// @link main _04_


    float _04_getDepth(int index) { return 0.5; }

//// @link use/types _05_

struct _05_SolidVertex {
  vec4 position;
  vec4 color;
  vec2 uv;
};


//// @link use/view _06_

layout(set = VIEW_BINDGROUP, binding = VIEW_BINDING) uniform _06_ViewUniforms {
  mat4 projectionMatrix;
  mat4 viewMatrix;
  vec4 viewPosition;
  vec2 viewResolution;
  vec2 viewSize;
  float viewWorldDepth;
  float viewPixelRatio;
} _06_viewUniforms;

vec4 _06_worldToView(vec4 position) {
  return _06_viewUniforms.viewMatrix * position;
}

vec4 _06_viewToClip(vec4 position) {
  return _06_viewUniforms.projectionMatrix * position;
}

vec4 _06_worldToClip(vec4 position) {
  return _06_viewToClip(_06_worldToView(position));
}

float _06_getPerspectiveScale(float w, float f) {
  mat4 m = _06_viewUniforms.projectionMatrix;
  float worldScale = m[1][1] * _06_viewUniforms.viewWorldDepth;
  float clipScale = mix(1.0, worldScale / w, f);
  float pixelScale = clipScale * _06_viewUniforms.viewPixelRatio;
  return pixelScale;
}

//// @link geometry/quad _07_

const ivec2 _07_QUAD[] = {
  ivec2(0, 0),
  ivec2(1, 0),
  ivec2(0, 1),
  ivec2(1, 1),
};

ivec2 _07_getQuadIndex(int vertex) {
  return _07_QUAD[vertex];
}

vec2 _07_getQuadUV(int vertex) {
  return vec2(_07_getQuadIndex(vertex));
}

//// @link getQuadVertex

vec4 _01_getPosition(int);
vec4 _02_getColor(int);
vec2 _03_getSize(int);
float _04_getDepth(int);

_05_SolidVertex getQuadVertex(int vertexIndex, int instanceIndex) {
  vec4 position = _01_getPosition(instanceIndex);
  vec4 color = _02_getColor(instanceIndex);
  vec2 size = _03_getSize(instanceIndex);
  float depth = _04_getDepth(instanceIndex);

  vec4 center = _06_worldToClip(position);

  vec2 uv = _07_getQuadUV(vertexIndex);
  vec2 xy = uv * 2.0 - 1.0;

  // Lerp between fixed size and full perspective.
  float pixelScale = _06_getPerspectiveScale(center.w, depth);
  size *= pixelScale;

  #ifdef HAS_EDGE_BLEED
  xy = xy * (size + 0.5) / size;
  uv = xy * .5 + .5;
  #endif

  center.xy += xy * size * _06_viewUniforms.viewResolution * center.w;

  return _05_SolidVertex(
    center,
    color,
    uv
  );
}"
`;

exports[`link links same module twice with different entry point 1`] = `
"#version 450
//// @link sub _01_


    float _01_used() { return 1.0; }

    vec4 _01_getPosition(int index) { return vec4(_01_used(), 0.0, 1.0, 1.0); }

    vec4 _01_getColor(int index) { return vec4(_01_used(), 0.0, 1.0, 1.0); }

//// @link main


    vec4 _01_getPosition(int index);
    vec4 _01_getColor(int index);
    void main() {
      vec4 a = _01_getPosition(0);
      vec4 b = _01_getColor(0);
    }
    "
`;

exports[`link tree shakes around identifiers 1`] = `
"#version 450
//// @link sub _01_


    float _01_used() { return 1.0; }

    vec4 _01_getPosition(int index) { return vec4(_01_used(), 0.0, 1.0, 1.0); }


//// @link main


    vec4 _01_getPosition(int index);
    void main() {
      vec4 a = _01_getPosition(0);
    }
    "
`;

exports[`link tree shakes around identifiers 2`] = `
"#version 450
//// @link sub _01_


    float _01_used() { return 1.0; }

    vec4 _01_getPosition(int index) { return vec4(_01_used(), 0.0, 1.0, 1.0); }


//// @link main


    vec4 _01_getPosition(int index);
    void main() {
      vec4 a = _01_getPosition(0);
    }
    "
`;

exports[`link tree shakes constants 1`] = `
"#version 450
//// @link sub _01_


    const vec4 _01_colorUsed = vec4(0.0, 0.1, 0.2, 0.0);
    vec4 _01_getColor() {
      return _01_colorUsed;
    }

//// @link main


    vec4 _01_getColor();
    void main() {
      vec4 a = _01_getColor();
    }
    "
`;

exports[`link tree shakes constants 2`] = `
"#version 450
//// @link sub _01_


    const vec4 _01_colorUsed = vec4(0.0, 0.1, 0.2, 0.0);
    vec4 _01_getColor() {
      return _01_colorUsed;
    }

//// @link main


    vec4 _01_getColor();
    void main() {
      vec4 a = _01_getColor();
    }
    "
`;
